### PR TITLE
TestMethodShouldContainAssertion (S2699) NUnit return value

### LINKED ISSUE #3751: S2699: should be skipped on tests with TestCaseSourceAttribute and return values
### Description

Rule ID: S2699

It does not respect `NUnit` `TestCaseSourceAttribute` for methods with return value.

### Repro steps
1. Reference NUnit 3;
2. Run following code:
```
[TestFixture]
internal class Equals_Should
{
  private static readonly TestCaseData[] Cases =
  {
    new TestCaseData("foo", "bar").Returns(false),
    new TestCaseData(100, 100).Returns(true),
  };

  [Test]
  [TestCaseSource(nameof(Cases))]
  public bool Test(object obj1, object obj2)
  {
    return obj1.Equals(obj2);
  }
}
```

### Expected behavior

No warnings as return value is checked by NUnit.

### Actual behavior

```
Warning	S2699	Add at least one assertion to this test case
```

### Known workarounds

No workarounds was found.

### Related information

* C# 8
* Visual Studio 2019 Pro. Version 16.7.6
* donetcore3.1
* Windows 10. Version 20H2.

### PR DESCRIPTION
**Pull request by alexandersorokin**
TestMethodShouldContainAssertion (S2699) rule takes into account a return value and the TestCaseSourceAttribute of NUnit test methods

Fixes #3751

### CODE REVIEW

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]
Hi @alexandersorokin, this looks good. There are just two small comments related to codding style conventions.

File: analyzers/src/SonarAnalyzer.Common/Helpers/UnitTestHelper.cs [hunk 1]
```diff
@@ -108,7 +108,18 @@ internal static class UnitTestHelper
             method.GetAttributes().Any(a => a.AttributeClass.IsAny(KnownExpectedExceptionAttributes));
 
         public static bool HasAssertionInAttribute(this IMethodSymbol method) =>
-            method.GetAttributes().Any(IsTestCaseAttributeWithExpectedResult);
+            !NoExpectedResultTestMethodReturnTypes.Any(method.ReturnType.Is) &&
+            method.GetAttributes().Any(IsAnyTestCaseAttributeWithExpectedResult);
+
+        private static readonly ImmutableArray<KnownType> NoExpectedResultTestMethodReturnTypes =
```
[the remark this answers is not in the document]
Comment (alexandersorokin) [reply analyzers/src/SonarAnalyzer.Common/Helpers/UnitTestHelper.cs]: fixed

File: analyzers/src/SonarAnalyzer.Common/Helpers/UnitTestHelper.cs [hunk 2]
```diff
@@ -108,7 +108,18 @@ internal static class UnitTestHelper
...
+            ImmutableArray.Create(
+                KnownType.Void,
+                KnownType.System_Threading_Tasks_Task
+            );
+
+        private static bool IsAnyTestCaseAttributeWithExpectedResult(AttributeData a) =>
+            IsTestCaseAttributeWithExpectedResult(a) ||
```
[the remark this answers is not in the document]
Comment (alexandersorokin) [reply analyzers/src/SonarAnalyzer.Common/Helpers/UnitTestHelper.cs]: fixed

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM! Thanks for the contribution!

**Pull request merged**